torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -352,6 +353,134 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+TensorView* TensorView::welfordRfactorHelper(
+    TensorView* tv,
+    const std::vector<int>& axes) {
+  // Hack:
+  // Semantically we should always keep the outputs of welfordOp scheduled
+  // the same but the user end cannot guarantee that.
+  // In order to guarantee that the rFactor is defined meaningfully the
+  // scheduling of the output TV that got the rfactor call is force replayed
+  // towards the other two
+
+  if (!sameAs(tv)) {
+    auto root = tv->getRootDomain();
+    auto this_root = getRootDomain();
+
+    // construct a trivial root domain map
+    std::unordered_map<IterDomain*, IterDomain*> id_map;
+    for (size_t i = 0; i < root.size(); i++) {
+      id_map[this_root[i]] = root[i];
+    }
+
+    // replay on the target tv
+    ReplayTransformations replay(domain()->domain(), id_map);
+
+    // construct the new tensor domain
+    std::vector<IterDomain*> new_id;
+    for (auto id : domain()->domain()) {
+      TORCH_INTERNAL_ASSERT(
+          replay.getReplay().count(id), "Welford Replay Failed");
+      new_id.push_back(replay.getReplay().at(id));
+    }
+
+    std::vector<bool> new_contig(
+        tv->domain()->contiguity().begin(), tv->domain()->contiguity().end());
```

Comment:
Contiguity should be based on the root domain right? I don't think contiguity should change, should just grab the one in the original domain.

Reply:
Thanks for pointing this out. I thought this was the only contiguity info around in a `tensordomain`, and it didn't look like root domain kept another version of it:
[https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/ir_internal_nodes.h#L501-L506](https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/ir_internal_nodes.h#L501-L506
)

I just needed to make a copy of the contiguity for constructing the new domain. Please let me know if there was another place to populate this vector.